elements/lisk-p2p/src/peer_server/peer_server.ts

```diff
@@ -0,0 +1,460 @@
+					},
+			  }
+			: {
+					sharedState: {
+						...restOfQueryObject,
+						...queryOptions,
+						height: queryObject.height ? +queryObject.height : 0, // TODO: Remove the usage of height for choosing among peers having same ipAddress, instead use productivity and reputation
+						protocolVersion: queryObject.protocolVersion,
+					},
+					internalState: {
+						...assignInternalInfo(
+							{
+								peerId,
+								ipAddress: socket.remoteAddress,
+								wsPort: remoteWSPort,
+							},
+							this._secret,
+						),
+						advertiseAddress: advertiseAddress !== 'false',
+						connectionKind: ConnectionKind.INBOUND,
+					},
+					peerId,
+					ipAddress: socket.remoteAddress,
+					wsPort: remoteWSPort,
+			  };
+
+		try {
+			validatePeerInfo(
+				incomingPeerInfo,
+				this._maxPeerInfoSize
+					? this._maxPeerInfoSize
+					: DEFAULT_MAX_PEER_INFO_SIZE,
+			);
+		} catch (error) {
+			this._disconnectSocketDueToFailedHandshake(
+				socket,
+				INCOMPATIBLE_PEER_INFO_CODE,
+				error,
+			);
+		}
+
+		const { success, error } = this._peerHandshakeCheck(
+			incomingPeerInfo,
+			this._nodeInfo,
+		);
+
+		if (!success) {
+			const errorReason = error || INCOMPATIBLE_PEER_UNKNOWN_REASON;
+
+			this._disconnectSocketDueToFailedHandshake(
+				socket,
+				INCOMPATIBLE_PEER_CODE,
+				errorReason,
+			);
+
+			return;
+		}
+
+		const incomingPeerConnection: IncomingPeerConnection = {
+			peerInfo: incomingPeerInfo,
+			socket,
+		};
+
+		this.emit(EVENT_NEW_INBOUND_PEER_CONNECTION, incomingPeerConnection);
+
+		return;
+	}
+
+	private _bindInvalidControlFrameEvents(socket: SCServerSocket): void {
+		// Terminate the connection the moment it receive ping frame
+		(socket as any).socket.on('ping', () => {
+			this._terminateIncomingSocket(
+				socket,
+				`Terminated connection peer: ${socket.remoteAddress}, reason: malicious ping control frames`,
+				true,
+			);
+		});
+		// Terminate the connection the moment it receive pong frame
+		(socket as any).socket.on('pong', () => {
+			this._terminateIncomingSocket(
+				socket,
+				`Terminated connection peer: ${socket.remoteAddress}, reason: malicious pong control frames`,
+				true,
+			);
+		});
+	}
+
+	private _handleIncomingPayload(ws: any, _req: any): void {
+		ws.on('message', (message: any) => {
+			// Pong message
+			if (message === '#2') {
+				return;
+			}
+
+			const MAX_EVENT_NAME_LENGTH = 128;
+
+			const peerIpAddress = ws._socket._peername.address;
+
+			const peerId = `${peerIpAddress}:${ws._socket._peername.port}`;
+
+			try {
+				const parsed = JSON.parse(message);
+
+				const invalidEvents: Set<string> = new Set([
+					'#authenticate',
+					'#removeAuthToken',
+					'#subscribe',
+					'#unsubscribe',
+					'#publish',
+				]);
+
+				if (
+					(parsed.event && typeof parsed.event !== 'string') ||
+					invalidEvents.has(parsed.event) ||
+					parsed.event.length > MAX_EVENT_NAME_LENGTH
+				) {
+					throw new Error('Received invalid payload');
```

Comment:
`throw new Error('Received invalid payload', parsed);` this way we could know the context better while debugging